### ISSUE TITLE
LoadBalancer creation retry happens every second

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#21630]
From customer logs, it looks like we retry to create a LoadBalancer after a failure every second.  This causes rate-limiting problems on AWS (and probably will on GCE also?).  We should have a less aggressive retry policy.

Related #21625

**Comment by bprashanth** [comment]
Isn't this a dupe of: https://github.com/kubernetes/kubernetes/issues/20316? 
I agree we should rate limit on failure with exponential backoff. We could possibly use the method kubelet uses to backoff image pulling. If you agress lets just close this as a dupe and review https://github.com/kubernetes/kubernetes/pull/21137 ?

**Comment by justinsb** [comment]
Sadly this is different.  This issue is that servicecontroller retries too quickly, without any real backoff, when a load balancer creation files.  Not specific to AWS - indeed I don't think the cloudprovider can implement the delay logic, because of issues discussed in #21625.

#20316 is specific to AWS, and not particular to load balancers, but the fix there is that AWS API calls can be delayed across the board when we get the "rate limit exceeded" error i.e. #21137 or #19335.  I think we should merge #19335 into 1.2 (i.e. asap); I've been doing my testing on an integration branch with it present.

I've proposed that we close #20316 as a dupe between this issue and #12121

**Comment by thockin** [comment]
I agree that the controller should backoff, regardless of the cloud
provider.

**Comment by thockin** [comment]
@kubernetes/goog-cluster @kubernetes/goog-control-plane anyone have time?

**Comment by quinton-hoole** [comment]
@thockin I can potentially take a look at this later this week, once ubernetes-lite tests are stabilized and docs are written.  If anyone has time before then, please have at it.  I'll update this issue before I start working on it, to avoid any confusion.  Provisionally assigning to myself.

**Comment by bprashanth** [comment]
@quinton-hoole we should just do https://github.com/kubernetes/kubernetes/issues/15095.

I wanted to add backoff to the workqueue (https://github.com/kubernetes/kubernetes/tree/b9cfab87e33ea649bdd13a1bd243c502d76e5d22/pkg/util/workqueue) when writing the RC manager but it got dropped. Essentially it would be great if we requeue items with non-terminal errors, rate limit retries and avoid backoff for transient flakes (or resource version conflict). 

A GREAT first step for the service controller would be re-writing it so it follows the RC manager pattern, it would get a lot of stuff for free. 

This specific issue will be mitigated by https://github.com/kubernetes/kubernetes/pull/19335

**Comment by quinton-hoole** [comment]
@bprashanth Regarding #15095 I'm not a fan of having such different polling frequencies configured for test vs production (as per the original proposal) as it will undoubtably mask errors in testing.  However I'm a huge fan of generalized backoff and rate limits, as proposed further down there.

It looks like #19335 (already merged) is good enough to remove this from the blockers queue for 1.2.  Anyone disagree?  ( @justinsb , @thockin )

**Comment by bprashanth** [comment]
Yeah I'm talking about unified backoff/retry policy in the workqueue. 
This will probably still happen on GCE. I suggest applying the same pattern we used for AWS backoff: https://github.com/kubernetes/kubernetes/pull/19335#discussion_r53561760

**Comment by bprashanth** [comment]
The link is broken. I meant encapsulate https://github.com/kubernetes/kubernetes/pull/19335/files#diff-fe3eb574eb9c040028e842395d85585dR92 into a Backoffer interface and reuse in GCE. You'll just need to check the cloudprovider specific error for backoff in `ReportError`.

**Comment by quinton-hoole** [comment]
@bprashanth Sounds perfectly sensible, but not a 1.2 blocker, right?

**Comment by justinsb** [comment]
I don't think we should rely on #19335.  It is a "last resort" for unknown problems, whose job it is to delay calls so we don't lock out the account.  I don't expect k8s to function well if it is triggered.  Also, #19335 is reactive, not preventative - it only triggers once we're already hitting the limit.

We know we have an over-eager retry loop here, and I think we should fix it.

I don't know that we need to fix it in 1.2, though I'm glad we're discussing it.

**Comment by quinton-hoole** [comment]
@justinsb Totally agreed.  Provisionally dropping to milestone next-candidate, but leaving P1, as I agree that it's important that we fix this.

**Comment by a-robinson** [comment]
Are you able to share any of the relevant logs? Even just the timestamps and code locations would be useful if the messages themselves have to be anonymized. I'd expect a retry no more than every 5 seconds for each service.

That policy is pretty poor itself and should be improved for next-candidate, but is 5x less bad than retrying every second. This is another place where getting the service controller in line with our other components would be very nice, as @bprashanth suggests.

**Comment by justinsb** [comment]
I redacted all the sensitive stuff.  There is only one LB here, with a name of a6<redacted>

```
E0220 21:00:26.205601       1 servicecontroller.go:187] Failed to process service delta. Retrying: Failed to create load balancer for service kube-system/kubedash: InvalidConfigurationRequest: ELB cannot be attached to multiple subnets in the same AZ.
    status code: 409, request id: <redacted>
I0220 21:00:26.789398       1 aws_loadbalancer.go:50] Creating load balancer with name: a6<redacted>
E0220 21:00:27.090195       1 servicecontroller.go:187] Failed to process service delta. Retrying: Failed to create load balancer for service kube-system/kubedash: Throttling: Rate exceeded
    status code: 400, request id: <redacted>
I0220 21:00:27.743317       1 aws_loadbalancer.go:50] Creating load balancer with name: a6<redacted>
E0220 21:00:28.113412       1 servicecontroller.go:187] Failed to process service delta. Retrying: Failed to create load balancer for service kube-system/kubedash: Throttling: Rate exceeded
    status code: 400, request id: <redacted>
I0220 21:00:28.697279       1 aws_loadbalancer.go:50] Creating load balancer with name: a6<redacted>
E0220 21:00:29.036503       1 servicecontroller.go:187] Failed to process service delta. Retrying: Failed to create load balancer for service kube-system/kubedash: Throttling: Rate exceeded
```

I agree that it is every 5 seconds in the code though.

I think what is happening is we have multiple changes here all being retried: #21952

With a relatively small fix proposed in #21982

**Comment by a-robinson** [comment]
> I think what is happening is we have multiple changes here all being retried: #21952

I agree, that seems likely. I'll take a look at #21982

**Comment by quinton-hoole** [comment]
@a-robinson Can this one be closed now that #21982 is merged?

**Comment by a-robinson** [comment]
Yup.
